contracts/core/BorrowableMarginPool.sol

```diff
@@ -0,0 +1,288 @@
+    event Repay(
+        address indexed oToken,
+        address indexed collateralAsset,
+        uint256 amount,
+        address indexed borrower,
+        address repayer
+    );
+
+    /**
+     * @notice check if the sender is whitelisted
+     */
+    modifier onlyWhitelistedBorrower() {
+        require(whitelistedBorrower[msg.sender], "MarginPool: Sender is not whitelisted borrower");
+
+        _;
+    }
+
+    /**
+     * @notice check if a borrower is whitelisted
+     * @param _borrower address of the borrower
+     * @return boolean, True if the borrower is whitelisted
+     */
+    function isWhitelistedBorrower(address _borrower) external view returns (bool) {
+        return whitelistedBorrower[_borrower];
+    }
+
+    /**
+     * @notice check if a oToken buyer is whitelisted
+     * @param _oTokenBuyer address of the oToken buyer
+     * @return boolean, True if the oToken buyer is whitelisted
+     */
+    function isWhitelistedOTokenBuyer(address _oTokenBuyer) external view returns (bool) {
+        return whitelistedOTokenBuyer[_oTokenBuyer];
+    }
+
+    /**
+     * @notice check if a options vault is whitelisted
+     * @param _optionsVault address of the options vault
+     * @return boolean, True if the options vault is whitelisted
+     */
+    function isWhitelistedOptionsVault(address _optionsVault) external view returns (bool) {
+        return whitelistedOptionsVault[_optionsVault];
+    }
+
+    /**
+     * @notice Set borrower whitelist status
+     * @param _borrower address of the borrower (100% of the time verified market makers)
+     * @param _whitelisted bool of whether it is whitelisted or blacklisted
+     */
+    function setBorrowerWhitelistedStatus(address _borrower, bool _whitelisted) external onlyOwner {
+        require(_borrower != address(0), "!_borrower");
+
+        whitelistedBorrower[_borrower] = _whitelisted;
+        emit SetBorrowWhitelist(_borrower, _whitelisted);
+    }
+
+    /**
+     * @notice Set options vault whitelist status
+     * @param _optionsVault address of the options vault
+     * @param _whitelisted bool of whether it is whitelisted or blacklisted
+     */
+    function setOptionsVaultWhitelistedStatus(address _optionsVault, bool _whitelisted) external onlyOwner {
+        require(_optionsVault != address(0), "!_optionsVault");
+
+        whitelistedOptionsVault[_optionsVault] = _whitelisted;
+        emit SetOptionsVaultWhitelist(_optionsVault, _whitelisted);
+    }
+
+    /**
+     * @notice Set oToken buyer whitelist status
+     * @param _oTokenBuyer address of the oToken buyer
+     * @param _whitelisted bool of whether it is whitelisted or blacklisted
+     */
+    function setOTokenBuyerWhitelistedStatus(address _oTokenBuyer, bool _whitelisted) external onlyOwner {
+        require(_oTokenBuyer != address(0), "!_buyer");
+
+        whitelistedOTokenBuyer[_oTokenBuyer] = _whitelisted;
+        emit SetOTokenBuyerWhitelist(_oTokenBuyer, _whitelisted);
+    }
+
```

Comment:
Just thinking loud here, if it make sense to keep this here or move it somehow to the Whitelist module

Reply:
Yeah I think moving this to Whitelist.sol make sense

Reply:
In principle I agree, but this would require a redeployment of Whitelist.sol with all of its existing whitelist state right? Think it makes sense to keep it in borrowable pool entirely to minimize the blast radius of this when deploying to prod.